src/home/homectl-pkcs11.c

```diff
@@ -171,27 +170,9 @@ int identity_add_pkcs11_key_data(JsonVariant **v, const char *uri) {
         if (r < 0)
                 return r;
 
-        pkey = X509_get0_pubkey(cert);
-        if (!pkey)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to extract public key from X.509 certificate.");
-
-        r = rsa_pkey_to_suitable_key_size(pkey, &decrypted_key_size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to extract RSA key size from X509 certificate.");
-
-        log_debug("Generating %zu bytes random key.", decrypted_key_size);
-
-        decrypted_key = malloc(decrypted_key_size);
-        if (!decrypted_key)
-                return log_oom();
-
-        r = crypto_random_bytes(decrypted_key, decrypted_key_size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to generate random key: %m");
-
-        r = rsa_encrypt_bytes(pkey, decrypted_key, decrypted_key_size, &encrypted_key, &encrypted_key_size);
+        r = X509_certificate_generate_volume_key(cert, &decrypted_key, &decrypted_key_size, &savedata, &savedata_size);
         if (r < 0)
-                return log_error_errno(r, "Failed to encrypt key: %m");
+                return log_error_errno(r, "Failed to generate secret key for X.509 certifcate: %m");
```

Comment:
your new function already logs on all errors internally. hence please drop the additional logging here.

in systemd we have this rule that each error should be logged about exactly once. for that you have to decide whether your function is of the logging kind (typically "main-program-like" functions) or of the non-logging kind (typically more "library-like"). If the former it should log about errors on its own, but still return the error code. If the latter it should leave the logging to the caller, and return the error code for that.

here it makes sense that X509_certificate_generate_volume_key() logs on its own, hence remove the logging here.